highstreets/features/processing_functions.py

```diff
@@ -0,0 +1,154 @@
+        "M"
+    )
+    hs_compare = hs.loc[(hs["month_year"]).isin(date_range)]
+    hs_compare = hs_compare.groupby(["highstreet_name"]).mean()
+
+    # merge dataframe
+    merged_df = pd.merge(hs_latest, hs_compare, on="highstreet_name")
+    # condtional labelling, if latest value not at least at average level, label at risk
+    merged_df["proportion"] = (
+        merged_df["txn_amt_wd_retail_x"] / merged_df["txn_amt_wd_retail_y"]
+    )
+    merged_df["labels"] = pd.qcut(
+        merged_df["txn_amt_wd_retail_x"] / merged_df["txn_amt_wd_retail_y"],
+        4,
+        labels=[0, 1, 2, 3],
+    )
+    merged_df = merged_df.reset_index()
+    # subset relevant columns
+    merged_df = merged_df[["labels"]]
+    return merged_df
+
+
+# loop and select data for relevant months
+def create_mean_sd_o2(highstreet_df, predictor_days):
+    df = []
+    for days in predictor_days:
+        df.append(highstreet_df.loc[(highstreet_df["count_date"] == days)])
+    df = pd.concat(df)
+
+    # group to create mean and std across all months
+    mean = df.groupby(["highstreet_name"]).mean().reset_index()
+    # rename and merge so less possibility of mistakes
+    mean = mean.rename({"h13": "mean_o2"}, axis=1)
+    std = df.groupby(["highstreet_name"]).std().reset_index()
+    std = std.rename({"h13": "std_o2"}, axis=1)
+    # merge and select columns
+    features = pd.merge(mean, std, on="highstreet_name")[
+        ["highstreet_name", "mean_o2", "std_o2"]
+    ]
+    return features
+
+    # loop and select data for relevant months
+
+
+def create_mean_sd_mcard(highstreet_df, predictor_months):
+    df = []
+    for months in predictor_months:
+        df.append(highstreet_df.loc[(highstreet_df["month_year"] == months)])
+    df = pd.concat(df)
+
+    # group to create mean and std across all months
+    mean = df.groupby(["highstreet_name"]).mean().reset_index()
+    # rename and merge so less possibility of mistakes
+    mean = mean.rename({"txn_amt_wd_retail": "mean"}, axis=1)
+    std = df.groupby(["highstreet_name"]).std().reset_index()
+    std = std.rename({"txn_amt_wd_retail": "std"}, axis=1)
+    # merge and select columns
+    features = pd.merge(mean, std, on="highstreet_name")[
+        ["highstreet_name", "mean", "std"]
+    ]
+    return features
+
+
+def create_gradient(hs, months):
+    """Use linear regression to calculate gradient of each high street
+    over the specified range of months"""
+    df = []
+    # create df of 3 months
+    for month in months:
+        df.append(hs.loc[(hs["month_year"] == month)])
+    df = pd.concat(df)
+    # craete ordinal time data
+    df["date_ordinal"] = pd.to_datetime(df["week_start"]).apply(
+        lambda date: date.toordinal()
+    )
+
+    # create list of df grouped by highstreet name, ordered
+    df_grad = []
+    names = []
+    for highstreet, subdf in df.groupby(["highstreet_name"], sort=True):
```

Comment:
This is a much cleaner way of extracting the gradients than what I had done! V good, uses pandas nicely. I was doing some awkward extracting of an array of all the data for each period I wanted gradients and looping over rows, this method, grouping by highstreet, is much better